File: lc/coreprep/implementation/main_reference_implementation/BinaryTreeDynamicNode.java
import java.util.*;
import java.util.function.*;

public class BinaryTreeDynamicNode {
  static class BinaryTree<T> {
    static class Node<T> {
      T data;
      Node<T> left;
      Node<T> right;

      Node(T data) {
        this.data = data;
        this.left = this.right = null;
      }
    }

    private Node<T> root;

    public BinaryTree() { this.root = null; }

    public void read(String[] nodes, Function<String, T> parser) {
      this.root = read(nodes, parser, 0);
    }

    private Node<T> read(String[] nodes, Function<String, T> parser, int idx) {
      if (nodes[idx].equals("null")) {
        return null;
      }

      Node<T> node = new Node<>(parser.apply(nodes[idx]));

      if (2 * idx + 1 < nodes.length) {
        node.left = read(nodes, parser, 2 * idx + 1);
      }

      if (2 * idx + 2 < nodes.length) {
        node.right = read(nodes, parser, 2 * idx + 2);
      }

      return node;
    }

    public void preOrder() {
      preOrder(this.root);
      System.out.println();
    }

    private void preOrder(Node<T> node) {
      if (node == null) {
        return;
      }

      System.out.print(node.data + " ");
      preOrder(node.left);
      preOrder(node.right);
    }

    public void preOrderIter() {
      Stack<Node<T>> st = new Stack<>();
      st.push(this.root);

      while (!st.empty()) {
        Node<T> node = st.pop();

        System.out.print(node.data + " ");

        if (node.right != null) {
          st.push(node.right);
        }

        if (node.left != null) {
          st.push(node.left);
        }
      }

      System.out.println();
    }

    public void inOrder() {
      inOrder(this.root);
      System.out.println();
    }

    private void inOrder(Node<T> node) {
      if (node == null) {
        return;
      }

      inOrder(node.left);
      System.out.print(node.data + " ");
      inOrder(node.right);
    }

    public void inOrderIter() {
      Stack<Node<T>> st = new Stack<>();
      Node<T> node = this.root;

      while (node != null || !st.empty()) {
        while (node != null) {
          st.push(node);
          node = node.left;
        }

        if (!st.empty()) {
          node = st.pop();
          System.out.print(node.data + " ");
          node = node.right;
        }
      }

      System.out.println();
    }

    public void postOrder() {
      postOrder(this.root);
      System.out.println();
    }

    private void postOrder(Node<T> node) {
      if (node == null) {
        return;
      }

      postOrder(node.left);
      postOrder(node.right);
      System.out.print(node.data + " ");
    }

    public void postOrderIter() {
      Stack<Node<T>> st = new Stack<>();
      Stack<Node<T>> processed = new Stack<>();

      st.push(this.root);
      while (!st.empty()) {
        Node<T> node = st.pop();

        processed.push(node);

        if (node.left != null) {
          st.push(node.left);
        }

        if (node.right != null) {
          st.push(node.right);
        }
      }

      while (!processed.empty()) {
        Node<T> node = processed.pop();
        System.out.print(node.data + " ");
      }
      System.out.println();
    }

    public void bfs() {
      Queue<Node<T>> q = new ArrayDeque<>();
      q.add(this.root);

      while (!q.isEmpty()) {
        Node<T> node = q.poll();

        System.out.print(node.data + " ");

        if (node.left != null) {
          q.add(node.left);
        }

        if (node.right != null) {
          q.add(node.right);
        }
      }
      System.out.println();
    }

    private int height() { return height(this.root); }

    private int height(Node<T> node) {
      if (node == null) {
        return 0;
      }
      return 1 + Math.max(height(node.left), height(node.right));
    }

    public void bfsRec() {
      int h = height();
      for (int i = 0; i < h; i++) {
        bfsRec(this.root, i);
      }
      System.out.println();
    }

    private void bfsRec(Node<T> node, int level) {
      if (node == null) {
        return;
      }

      if (level == 0) {
        System.out.print(node.data + " ");
      } else {
        bfsRec(node.left, level - 1);
        bfsRec(node.right, level - 1);
      }
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      BinaryTree<Integer> tree = new BinaryTree<>();
      String[] nodes = in.nextLine().trim().split(" ");

      tree.read(nodes, Integer::parseInt);

      tree.preOrder();
      tree.preOrderIter();

      tree.inOrder();
      tree.inOrderIter();

      tree.postOrder();
      tree.postOrderIter();

      tree.bfs();
      tree.bfsRec();
    }
  }
}

File: lc/coreprep/topics/dynamic_programming/binary_tree_dynamic_node.py
from collections import deque

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def build_tree(nodes):
    def build_tree_rec(nodes, curr_idx):
        if curr_idx >= len(nodes) or nodes[curr_idx] == "null":
            return None
        node = TreeNode(int(nodes[curr_idx]))
        node.left = build_tree_rec(nodes, 2 * curr_idx + 1)
        node.right = build_tree_rec(nodes, 2 * curr_idx + 2)
        return node
    return build_tree_rec(nodes, 0)

def pre_order(root):
    def pre_order_helper(root, nodes):
        if root is None:
            return
        nodes.append(root.val)
        pre_order_helper(root.left, nodes)
        pre_order_helper(root.right, nodes)

    nodes = []
    pre_order_helper(root, nodes)

    if len(nodes) != 0:
        for node in nodes:
            print(node, end=' ')
        print()

def pre_order_iter(root):
    nodes = []

    st = []
    st.append(root)

    while len(st) != 0:
        node = st.pop()
        nodes.append(node.val)

        if node.right:
            st.append(node.right)

        if node.left:
            st.append(node.left)

    if len(nodes) != 0:
        for node in nodes:
            print(node, end= ' ')
        print()

def in_order(root):
    def in_order_helper(root, nodes):
        if root is None:
            return

        in_order_helper(root.left, nodes)
        nodes.append(root.val)
        in_order_helper(root.right, nodes)

    nodes = []
    in_order_helper(root, nodes)

    if len(nodes) != 0:
        for node in nodes:
            print(node, end = ' ')
        print()

def in_order_iter(root):
    nodes = []

    st = []
    curr_node = root

    while len(st) != 0 or curr_node is not None:
        while curr_node is not None:
            st.append(curr_node)
            curr_node = curr_node.left

        if len(st) != 0:
            node = st.pop()
            nodes.append(node.val)

            if node.right:
                curr_node = node.right

    if len(nodes) != 0:
        for node in nodes:
            print(node, end = ' ')
        print()

def post_order(root):
    def post_order_helper(root, nodes):
        if root is None:
            return

        post_order_helper(root.left, nodes)
        post_order_helper(root.right, nodes)
        nodes.append(root.val)

    nodes = []
    post_order_helper(root, nodes)

    if len(nodes) != 0:
        for node in nodes:
            print(node, end = ' ')
        print()

def post_order_iter(root):
    nodes = []

    st = []
    rev_st = []
    st.append(root)

    while len(st) != 0:
        node = st.pop()

        rev_st.append(node.val)

        if node.left:
            st.append(node.left)

        if node.right:
            st.append(node.right)

    while len(rev_st) != 0:
        nodes.append(rev_st.pop())

    if len(nodes) != 0:
        for node in nodes:
            print(node, end =  ' ')
        print()


def bfs(root):
    nodes = []

    q = deque()
    q.append(root)

    while q:
        node = q.popleft()
        nodes.append(node.val)

        if node.left:
            q.append(node.left)

        if node.right:
            q.append(node.right)

    if nodes:
        for node in nodes:
            print(node, end = ' ')
        print()
    

def bfs_rec(root):
    def bfs_rec_helper(q, nodes):
        if not q:
            return

        node = q.popleft()
        nodes.append(node.val)

        if node.left:
            q.append(node.left)

        if node.right:
            q.append(node.right)

        bfs_rec_helper(q, nodes)


    nodes = []

    q = deque()
    q.append(root)

    bfs_rec_helper(q, nodes)

    if nodes:
        for node in nodes:
            print(node, end = ' ')
        print()

def main():
    nodes = input().split()

    root = build_tree(nodes)
    pre_order(root)
    pre_order_iter(root)

    in_order(root)
    in_order_iter(root)

    post_order(root)
    post_order_iter(root)

    bfs(root)
    bfs_rec(root)

if __name__ == "__main__":
    main()